### ISSUE TITLE
Support for deletion approval workflows

### ISSUE THREAD

**Issue by aetdr** [globaleaks/GlobaLeaks#3061]
Hi

Currently, a reciever with permission to delete submissions is free to wipe out a submissions whenever they want. 
There should be a way to control this by introducing some sort of approval workflows, so deletion can not be triggred unless at least 2 recievers have approved this action (four-eyes principle) 

It should be relatively easy to implement.

**Comment by evilaliv3** [comment]
Thank you for the feedback @aetdr.

As evolution of the Deletion Feature i consider we could improve the "Submission States" adding the possibility to configure a data retention policy after which deleting the submission.
This would make it possible to create custom states like a "To be deleted" state where to preserve the submission for a certain period before deletion.
This in combination with a feature to notify the change of state to other recipients would make it possibile for every recipient to change the state of the submission to avoid deletion.

What do you think?

**Comment by aetdr** [comment]
Hi @evilaliv3 

Can think of a couple possible issues with the approach:

-  If recipient1 triggers deletion, and nobody reacts to the mails sent in the data retention period, the submission will still get deleted. That should not happen. A submission should not get deleted before the second recipient actively confirms the request.

- In an emergency situation, when due different reasons, sensitive submission has to be removed from a public-facing server immediately, and where two recipients agree on it - they would still have to wait on the data retention period to expire. That should not happen, if two recipients agree, they should be able to delete the submission immediately.  

The way I vision it:

1.  A recipient with permission to delete can trigger deletion
    -   This action sets the submision in "deletion requested" state
    -   This action sends an email informing all the recipients with access to the submission about the state change
    -   Recipients notification settings should not prevent them from receiving this email, alternatively there should be a way to do this without requiring email
 

2.  A recipient different from the one who triggered the request can confirm the request
    - The confirmation could be a link in the mail, ideally requiring the user to login again, followed by prompt "Are you sure... YES/NO" 
    - When confirming recipient click on "YES", the submission is set in "deletion approved" state
    - This action sends an email to other recipients informing on state change. Again notification settings should not stay in the way or there should be an alternative way to inform recipients.
    - Maybe a simple account email check that verifies that requester email is not equal to confirmer email shoul be performed before state change


3. Any recipient with permission to delete, can immediately invoke deletion of any submission in "deletion approved" state 
   - Key information should be preserver in the audit log

**Comment by evilaliv3** [comment]
Thank you @aetdr for the feedback.

Lets see if we could get other feedback from other users.

We shall consider many situations and the solution should be simple and work like a charm by default.

I find your proposal interesting but i consider it could be significantly complicated and achieve opposite results. The success of the current implementation is in fact based on its simplicity and security by default. I understand what led you to propose this nice proposal, but maybe the current design and a proper audit log and a data retention of 14 days could work nicely and remain more simple?

In your proposal i see many issue like on conditions where:
- only one recipient is configured
- the solution should work when multiple recipients are configured but only one is the active one and the other are set just for backup reasons or just to stay formally in copy.

both these situations requires an automatic data retention policy to avoid maintaining data indefinitely.

\cc @elbill @schris-dk @larrykind @maxmois: Have you ever collected similar needs or suggestions from your users?

**Comment by larrykind** [comment]
Hi everyone, thank you for mentioning me. 
In my opinion this is an interensting feature, but I think a simplification can be achieved. 

- For evey context it could be optionally elected a "master" receiver, just one.
- Deletions (or better generally any state changes, excepted "open") shoud be approved by this master receiver. 
- A timeout should be introduced to perform the state change after it expires, even it is not approved.
- The notification email will be sent only to the master receiver with link to confirm the state change, in the email content will compare also timeouts to make this master receiver aware of the expiration.
- Agree with audit log tracking of the procedure.

Clearly the master receiver will perform deletion without anyone else permission. 
If no "master" is elected every reciver can perform state change without approval procedure.
I think this could fit also with new feature #2595 

@evilaliv3 - This feature maybe with #2595 could drive into a new and low-privileged receiver profile "collaborator" as suggested also in ANAC deliberation no 469 (june 2021)

**Comment by evilaliv3** [comment]
Thank you so much @larrykind, i consider that you feedback is very very valuable and  fully agree on your suggestion.

I revised your message to avoid using the word master; within our project project we prefer to use alternative inclusive language and terminology. To understand the reason and join our efforts i would suggest to read ideas of @mallory and @nllz.

I've some open questions:
- Shall we ask a motivation to the recipient requesting the deletion?
- Would it be useful to make it possible to configure multiple users that could authorize these operation.

@giorgiofraschini: what do you think?

I would appreciate if you all could look as well at ticket https://github.com/globaleaks/GlobaLeaks/issues/3056 that is actually strictly related. thank you!

**Comment by elbill** [comment]
@evilaliv3 and everybody, It is all a bit overwhelming. The simplest solution I can think off is one user to request deletion and if within 15 (or xx) days none has objected the report to be deleted.
An archival mechanism where reports will be not deleted after that but archived and any user will be able to bring back from archive (and this to be logged in the audit for GDPR compliance) is very appealing to me.
I don't discredit what @larrykind proposed, it is very useful, however my experience shows that unless you have a very organised team of recipients that do this on a large scale, "main" receiver concept can get complex and confusing.

**Comment by giorgiofraschini** [comment]
I see @larrykind's points at all but I definitely agree with @elbill here. It makes more sense in the large majority of projects I know of.

**Comment by larrykind** [comment]
@evilaliv3 yes in my opinion a motivation could be useful for deletion/state change request. I suggest to replicate the existing chat system in a new flavour for internal receivers discussions.
I don't see the need to have multiple main users that can perform state change authorization, I tried to propose to setup a kind of  gerarchy of receivers.
@elbill and @giorgiofraschini I think I understand your point, but  the Idea of main receiver was straight to "not organized" group of recipients (as the majority is in my experience), the **optional** need of a main one to approve their state change decisions.
I see more difficult to achieve the idea of multiple and collaborating receivers, they really must have a well organized team.

**Comment by elbill** [comment]
@larrykind could not agree more regarding "to replicate the existing chat system in a new flavour for internal receivers discussions"  #1957 . I would prioritise this over the support for deletion workflows.
For the rest of it I take your word :-)

**Comment by evilaliv3** [comment]
Thank you both for your feedback.

Fine for me to proceed with the latest proposal by @elbill that is actually in line with my original proposal.

So we would actually need:
- To mark a submission to be deleted adding a deleted state
- To make it possible to configure the data retention policy of the deleted state
- To notify recipients when the state is changed
- To renew the expiration date of a data retention policy if the submission is removed from the deleted state

**Comment by aetdr** [comment]
Helo @evilaliv3 and others

My earlier points are still valid:

- passivity does not equal confirmation (what if email notifications are swithched off, vacation, maternity leave, user left company)
- control is taken away from recipients (need to wait 14 days in order to delete - as an admin I will probably have to script a solution to help recipients and manipulate this if this gets enforced)

Perhaps my message was not clear:

- This feature should not be enabled for everyone by default (plenty of cases where it does not make sense). This is something you opt-in for as an admin, and by that you take responsibility for all those cases where it could go wrong.
- When I mentioned word "state" I did not mean "submission status" (as new/opened/closed)  but more like deletion process state.

Anyway, email notifications can not be trusted as an user can switch them off. 
The whole thing could be solved simply with currrent deletion button within the submission, a couple of clicks from different recipients and informative modals.

**Comment by aetdr** [comment]
Hi @evilaliv3 

An update on this... In absence of a better way to handle this situation, the way our organization wants to implement this is by removin access to deleting submissions from recipients. Admins should instead to do the deletion after the proces is initiated by a recipient.

Admins have that big red button, and can delete all submissions from a node, and that meets the requirement if only 1 submission is present.

If there are multiple submissions, where only one should be deleted, than we need to copy submission ID from the audit log, and execute a SQL query server-side. That does the job, but bypasses the intention and audit log if we do not add the entry ourselves. This is kind of messy.

I was wondering, if a feature that enables admins to delete single submissions via WebUI is a valid approach to handle this situation.

**Comment by evilaliv3** [comment]
@aetdr : Actually in the globaleaks threth model administrator are specifically not involved in these activities: https://docs.globaleaks.org/en/main/security/ThreatModel.html

Maybe you could give a single recipient the priviledge to delete submissions? Would this work in your use case?

**Comment by elbill** [comment]
Maybe what @aetdr workarround is to consider a recipient as "report admin" with rights to delete reports while others can not.

**Comment by aetdr** [comment]
Hi @evilaliv3 and @elbill 

Personally I agree. Ideally, GL admins should not get involved in this process.
The decision is not mine. There are some regulations involved and the 4 eyes principle is one of requirement. 
The way it is implemented is acceptable by those who enforce the regulation.

We have been looking into possibility to grant the deletion permission only to one recipient. Actually we split recipients in two pseudo-groups (primary recipients with permission to do everything except deletion, and secondary with less permissions). Originally the primary recipient was able to do deletion.

The concern is that a submission can be about the recipient themselves (as both the recipient and the whistleblower work in same organization). In that case the recipient is able to wipe the submission before anyone reads it. There will be some audit logs left, but the content is gone.
